### ISSUE TITLE
[FG3] LinkageErrors when trying to run with non-mod dependency

### ISSUE THREAD

**Issue by darkevilmac** [MinecraftForge/ForgeGradle#583]
I've tried posting in the forge discord for help but didn't end up getting a response. I don't really think this fits the forums because it *seems* like a regression more than anything else.

When updating my workspace from 1.12 to 1.13 I've run into an issue with one of my dependencies. Previously I was able to add the dependency to the project and reference it and run the client or server and everything worked fine. With 1.13 though I end up getting errors whenever my mod references a part of my dependency that calls any forge/vanilla code.

I'm not sure if I've messed something up on the dependency side or if the issue is with how I'm adding the dependency to my project. I've been stuck on this for a good while now and don't really know how to move forward.

This is one of the exceptions that gets thrown when I try and run the game: https://gist.github.com/darkevilmac/f36f489473be5c1d9697b41c860f045f
This the repo of the dependency I'm adding: https://github.com/TridentMC/Compound/
This is the repo of the mod that relies on the dependency: https://github.com/tridentmc/Movingworld

**Comment by mysticdrew** [comment]
I am having the same issue with a non mod dep.
Build file. https://gist.githubusercontent.com/mysticdrew/819c04ada701f2f2d220b689e34c42aa/raw/470634683929da0ca90194559bea0ac8f2be6279/build.gradle

`java.lang.LinkageError: loader constraint violation: when resolving method "journeymap.client.api.util.UIState.newInactive(Ljourneymap/client/api/display/Context$UI;Lnet/minecraft/client/Minecraft;)Ljourneymap/client/api/util/UIState;" the class loader (instance of cpw/mods/modlauncher/TransformingClassLoader) of the current class, journeymap/client/render/map/GridRenderer, and the class loader (instance of sun/misc/Launcher$AppClassLoader) for the method's defining class, journeymap/client/api/util/UIState, have different Class objects for the type net/minecraft/client/Minecraft used in the signature`

**Comment by darkevilmac** [comment]
I'm still having this issue as well, I'd honestly be happy with even just a workaround that I can use until there's a proper fix. I can't update any of my mods until I can actually use my libraries...

**Comment by mysticdrew** [comment]
Check the forge discord, I posted a workaround in the 1.13.2 channel.

**Comment by LexManos** [comment]
Mystics solution is not a good one you should not need to do it. And you should post any "solutions" here so others dont have to go searching. again try and trim down the use case and I'll look into it tomarrow when I'm back in town

**Comment by darkevilmac** [comment]
It's definitely hacky, I just need something I can use until this is fixed. I'd rather deal with using a hack and at least start to get my mods up to date than just be stuck waiting.

**Comment by mysticdrew** [comment]
@LexManos 
Sorry, I kind of forgot about this. Here is a simple testmod. 
https://github.com/mysticdrew/TestMod

I have it use the journeymap-api dep to cause the linkage error. 
https://github.com/TeamJM/journeymap-api

**Comment by mysticdrew** [comment]
So the testmod calls a class in the API that needs the Minecraft class. The api's classloader loads the class and then the linkage error happens.

This worked fine in 1.12.2. 

For reference since this was a while ago, here is the workaround that I mentioned earlier in this topic.

```java
ext {
    jmApiClassesDir = new File("build/journeymap-api")
}

minecraft{
    environment "MOD_CLASSES", [
            sourceSets.main.output.resourcesDir.absolutePath,
            sourceSets.main.output.classesDirs.asPath,
            project.jmApiClassesDir.absolutePath
    ].join(File.pathSeparator)
}

task extractApi(type: Copy) {
    project.jmApiClassesDir.mkdirs()
    from zipTree(project.configurations.includeInJar.find { it.name.startsWith("journeymap-api-") })
    into project.jmApiClassesDir
}

compileJava.dependsOn(extractApi)
```

**Comment by sciwhiz12** [comment]
As #758 is fixed which is very much related to this issue too, I am posting the solution found there and a more indepth explanation.

**TL;DR: Add the `FMLModType: LIBRARY` attribute to the non-mod dependency (or if not possible, shade it in).**
<hr>

### TransformingClassLoader and FMLModType

Minecraft Forge relies on the [modlauncher](https://github.com/cpw/modlauncher) library, which allows for "launching Java programs with in-flight transformation ability". This is done by having certain paths go through a `TransformingClassLoader`, which passes classes being loaded using it through transformers (The specifics of how classes are transformed is not revelant here). 
A [`TransformingClassLoader`](https://github.com/cpw/modlauncher/blob/3cf110cd0b2a5bcd8a44057fd0513e38a4b06f97/src/main/java/cpw/mods/modlauncher/TransformingClassLoader.java) must be configured beforehand (via the [`ILaunchHandlerService#configureTransformationClassLoader(ITransformingClassLoaderBuilder)`](https://github.com/cpw/modlauncher/blob/3cf110cd0b2a5bcd8a44057fd0513e38a4b06f97/src/main/java/cpw/mods/modlauncher/api/ILaunchHandlerService.java#L30) method) with the parent `Path`s where classes originate from for them to be passed through the transformers, otherwise the class is passed onto the parent class loader for regular loading.

FML/Forge introduces a special manifest attribute of `FMLModType` ([from ForgeSPI](https://github.com/MinecraftForge/ForgeSPI/blob/8ddb2cc7d9443e186f0cb383d94af266e4c2ee1f/src/main/java/net/minecraftforge/forgespi/locating/IModFile.java#L31-L33)), which is used to define if a JAR contains a mod, library, or language provider (`#MOD`, `#LIBRARY`, and `#LANGPROVIDER` respectively). For mods located via the FML mod locators (mods folder, classpath, exploded mod directory), the locators defaults the type to `MOD` if not present.
JARs marked with the `LIBRARY` mod type are [added to the common launch handler as additional libraries](https://github.com/MinecraftForge/MinecraftForge/blob/0239a736ec117dbe1956b814c1ef28b9af1888d5/src/fmllauncher/java/net/minecraftforge/fml/loading/FMLLoader.java#L218), which are then passed to the transforming class loader builder as additional `Path`s which are loaded and transformed.

### The LinkageError

The `LinkageError` from mysticdrew's comment at https://github.com/MinecraftForge/ForgeGradle/issues/583#issuecomment-493774075 describes the reason for the error in the message: a clash between the two class loaders. 

The `UIState` class comes from the journeymap-api JAR, which is not marked as `FMLModType: LIBRARY`, and is not a mod: this means that the class is passed by `TransformingClassLoader` to the parent class loader (in this case, the `Launcher$AppClassLoader` from the JDK). This means that when the JDK loaded the `UIState` class and the classes referenced from the method signatures (in this case, from `UIState#newInactive`), it loaded the `Minecraft` class on the `AppClassLoader`.

Then, another class from within a mod now tries to call the method `UIState#newInactive`. This class has it's parent class loader as the `TransformingClassLoader` (all mods JARs are put through the transformers). This means that the `Minecraft` class is loaded on the `TransformingClassLoader`. Now, when the mod clas calls the method, the JVM tries to check the method signature's classes with the classes of the objects being passed into it, sees that `Minecraft` class from the method signature is different from the `Minecraft` class of the object being passed into it, and subsequently errors out with the `LinkageError`.

_A bit of speculation/assumption:_ I assume that the reason why the JVM checks this is to prevent cross-classloader-caused errors. Since the class objects are different, then the JVM cannot ensure that the object's data actually conforms to the class being specified. (Imagine `A` being loaded on two different classloaders: one where it has the method `get()` and one where it didn't: trying to pass in an object from the second class without the `get()` to a method which expects an object from the first class with the `get()` would cause issues.)

### Solution

Adding the `FMLModType: LIBRARY` attribute to the non-mod dependency JAR solves the issue by also passing classes from that JAR to the `TransformingClassLoader` that is also used by mods, therefore the `Minecraft` class (when being resolved due to the `UIState#newInactive` class) is always loaded via the `TransformingClassLoader`, always resulting in the same class object. This is the solution that should be applied to the `journeymap-api` JAR (and it has, see #758).

Note that this problem only occurs in dependencies which do not contain mods, but reference classes which are transformed by the `TransformingClassLoader` (such as from the Minecraft codebase): if the dependency does not reference such classes, or if those referenced classes are not transformed, then no `LinkageError` occurs (because it is passed to the parent class loader, therefore having the same class object).
